### ISSUE TITLE
Warn on stderr when producing output that cannot be checked with --check

### ISSUE THREAD

**Issue by joveian** [BLAKE3-team/BLAKE3#281]
Currently per #33 and the --check [spec](https://github.com/BLAKE3-team/BLAKE3/blob/master/b3sum/what_does_check_do.md) b3sum --check rejects checking files with the unicode replacement character.  However, there is no warning or error when b3sum without --check produces output that can't be checked with --check, which seems less than ideal since checking files is usually the point of producing the list in the first place.  Since the spec seems to indicate that creating a list that can't be checked with --check is intended, I suggest a warning on stderr in that case.

**Comment by joveian** [comment]
Sorry for the noise, I messed up using GitHub and should have had the changes on a branch so I could make another branch for other patches.

**Comment by oconnor663** [comment]
> checking files is usually the point of producing the list in the first place

I'm not sure this is right. Personally, I never use `--check`, even though I use `b3sum` by hand frequently. My instinct is that we should optimize for the case where the human is reading `b3sum` output with their eyeballs.

That said, invalid unicode paths should also be pretty rare, so I don't think most people eyeball experience is going to be affected either way by what we decide here.

**Comment by oconnor663** [comment]
What do you think of doing a summarized warning at the bottom. I think that might be less spammy, and also more likely to make the user see it, but possibly more confusing about which file caused the issue. Maybe something like this?

```diff
diff --git a/b3sum/src/main.rs b/b3sum/src/main.rs
index 91a87a1..31a3c15 100644
--- a/b3sum/src/main.rs
+++ b/b3sum/src/main.rs
@@ -565,6 +565,8 @@ fn main() -> Result<()> {
     let thread_pool = thread_pool_builder.build()?;
     thread_pool.install(|| {
         let mut some_file_failed = false;
+        let mut found_invalid_unicode = false;
+        let mut found_literal_replacement_char = false;
         // Note that file_args automatically includes `-` if nothing is given.
         for path in &args.file_args {
             if args.check() {
@@ -585,6 +587,26 @@ fn main() -> Result<()> {
                     eprintln!("{}: {}: {}", NAME, path.to_string_lossy(), e);
                 }
             }
+            match path.to_str() {
+                Some(s) => {
+                    if s.contains('�') {
+                        found_literal_replacement_char = true;
+                    }
+                }
+                None => found_invalid_unicode = true,
+            };
+        }
+        if found_invalid_unicode {
+            eprintln!(
+                "{}: a filename contains invalid unicode, output is not checkable",
+                NAME,
+            );
+        }
+        if found_literal_replacement_char {
+            eprintln!(
+                "{}: a filename contains a literal � character, output is not checkable",
+                NAME,
+            );
         }
         std::process::exit(if some_file_failed { 1 } else { 0 });
     })
```

**Comment by oconnor663** [comment]
Since we might end up adding summary messages like this for both hashing (here) and checking (your other PR), it might also make sense to split out a separate helper function for each case, to clean up `main()` a little bit.

**Comment by joveian** [comment]
A summary message sounds good to me, although I think it should count the  number of issues for each type as well.  The example I found came from a Windows Romanian codepage, so for some people those cases may not be rare at all (that may just make b3sum unusable for them, but it would be nicer to give a summary in that case).  I'm not sure how common non-unicode compatible character sets are these days.

Also, I think the documentation should mention this issue more prominently than on the detailed --check page.  Would you accept a man page?

Splitting out the two cases and just returning the exit status makes sense to me.  If the other PR looks ok to you with the u64 change maybe committing that first would be easiest or else working on both changes in #285.

I'm using b3sum for backups so I'm hashing or checking many files (the speed of b3sum makes it great for this vs others; I make the check files with 'fd -t f -X b3sum --').  Even when checking one file I don't think it is a good idea to encourage comparing hashes by eye since that is slow and error prone.  I think it would make sense to have a third mode of operation where the arguments are alternating hashes to check and file names.  This would avoid the valid unicode limitation on file names and make it easier to do automated comparison with some random hash pulled from a webpage or such.  It doesn't look like any of the other similar utilities have such a feature currently (except that FreeBSD's sha256, etc. commands use -c to provide a single hash to check a given file against, conflicting with the more common check file).  I could try to implement that later if you would consider such an option.

**Comment by oconnor663** [comment]
#285 is landed! I put up 564ac5429ac95e30b27fbad4b1334c1ffeff240d on a branch as a candidate implementation for this issue.

> I'm using b3sum for backups

I added `--check` for compatibility, and because like you said it's better than checking by eye. That said, I've never been entirely comfortable with it as a security tool. The most glaring issue is that it does nothing to stop an attacker from _adding extra files_ to the backup directory. There are some use cases where this is no big deal, but if we're talking about backing up e.g. `~/.config` it's a huge security hole.

There's probably another issue or two where these questions have come up, but if I was going to try to design a really robust tool for verifying directory trees, I'd want to consider these features:

- Directory hashes, as opposed to just file hashes. This probably just means sorting the contents and hashing them, but even that requires some sort of formatting decision.
- Metadata. Does the difference between `chmod 644` and `chmod 755` show up in the hash? Timestamps? (Presumably not.) What about Unix-specific metadata like the sticky bit? Note that `tar` has many flags for controlling how much of this stuff gets included.
- Symlinks? This has been a sticking point between Windows and Unix for ever, and utilities like Git have had to introduce complicated hacks to work around the differences.

These sorts of questions are really important for a robust directory hashing tool, but they're pretty far out of scope for `b3sum`. It's not clear to me whether the right solution even looks like `b3sum` (i.e. hashing stuff that's already on the filesystem), vs some more opinionated archiver like `tar` or `zip` that could offer a "canonical" mode and actually make guarantees about what metadata is present.

**Comment by joveian** [comment]
Thanks!

I agree that avoiding feature creep is the right choice for b3sum in general, although an additional way to check hashes provided on the command line seems like it could be in scope to me.  But I also understand not wanting to add anything that the other similar utilites don't have.

I also use NetBSD's [mtree](https://man.netbsd.org/mtree.8) (via pkgsrc bootstrap when not on NetBSD, although I doubt it works on Windows outside the Linux subsystem) plus tar and the [scrypt utility](https://www.tarsnap.com/scrypt.html).   I have a large collection of music and games and I'm mostly looking to detect bitrot (just in case, although I haven't seen it for a while since drives do quite a bit of checking and tend to either work perfectly or not at all).  That stuff I keep outside archives since I sometimes want to acess something individually.  Creating a new b3sum checkfile of files outside encrypted archives and comparing to the old one from the encrypted archive detects any new files and makes sense since it is easy to do, although I'm not too worried about physical security in this case.  The thing that prompted me to switch was actually that I had been saving file permissions but made stuff read only after making the mtree file and there isn't an easy way to just check part of the mtree checkfile information at the moment.  While it is easy enough to not save permissions with mtree, that particular backup task seems like a good fit for b3sum which is much faster at the moment and one file per line also makes it easy to check individual files or directories.  I should also add blake2 and blake3 support to mtree and maybe a checkfile filter tool.

I'll argue again for keeping a count of files with invalid unicode and literal '�' characters just to give an idea of how big of an issue there is (grep '�' | wc -l should work too but since it is so easy to do it seems friendlier to just count and this would also match the failures warning).  

Let me know what you think in terms of documentation if you would accept a man page or if at least a more detailed --help message to describe the invalid unicode or literal '�' issue would be ok.
